# **Victoria Plumbing** License Details from Official Website

**Source:** `https://consumer.etoolbox.pic.vic.gov.au/_layouts/cc/pic_validatepp.aspx`

**Recommendation:** This code depends on the recent Web structure of Victoria Plumbing website licensing portal, if they will change their webpage's structure in future then this code won't work.

# **Step 2: `Data Ingestion`**

**Importing essential Libraries**

In [1]:
import selenium
import time
from selenium import webdriver

from bs4 import BeautifulSoup
import pandas as pd

In [2]:
from bs4 import BeautifulSoup
import pandas as pd

**Creating a List of inputs for Web portal search**

In [3]:
import string
char_one=list(string.ascii_uppercase)
char_two=char_one.copy()
print(char_one)
print(char_two)

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']


In [4]:
len(char_one)

26

**Final Input List:** Creating a list of letters using the two character lists from above, to get all the combinations of `First name` & `Last name`.

In [5]:
Letters=[char_one,char_two]
from itertools import product

Letters_comb=list(map(' '.join, product(*Letters)))
print(Letters_comb)

['A A', 'A B', 'A C', 'A D', 'A E', 'A F', 'A G', 'A H', 'A I', 'A J', 'A K', 'A L', 'A M', 'A N', 'A O', 'A P', 'A Q', 'A R', 'A S', 'A T', 'A U', 'A V', 'A W', 'A X', 'A Y', 'A Z', 'B A', 'B B', 'B C', 'B D', 'B E', 'B F', 'B G', 'B H', 'B I', 'B J', 'B K', 'B L', 'B M', 'B N', 'B O', 'B P', 'B Q', 'B R', 'B S', 'B T', 'B U', 'B V', 'B W', 'B X', 'B Y', 'B Z', 'C A', 'C B', 'C C', 'C D', 'C E', 'C F', 'C G', 'C H', 'C I', 'C J', 'C K', 'C L', 'C M', 'C N', 'C O', 'C P', 'C Q', 'C R', 'C S', 'C T', 'C U', 'C V', 'C W', 'C X', 'C Y', 'C Z', 'D A', 'D B', 'D C', 'D D', 'D E', 'D F', 'D G', 'D H', 'D I', 'D J', 'D K', 'D L', 'D M', 'D N', 'D O', 'D P', 'D Q', 'D R', 'D S', 'D T', 'D U', 'D V', 'D W', 'D X', 'D Y', 'D Z', 'E A', 'E B', 'E C', 'E D', 'E E', 'E F', 'E G', 'E H', 'E I', 'E J', 'E K', 'E L', 'E M', 'E N', 'E O', 'E P', 'E Q', 'E R', 'E S', 'E T', 'E U', 'E V', 'E W', 'E X', 'E Y', 'E Z', 'F A', 'F B', 'F C', 'F D', 'F E', 'F F', 'F G', 'F H', 'F I', 'F J', 'F K', 'F L', 'F M'

In [10]:
Letters_comb.index('A A')

0

In [7]:
Letters_comb[393:]

['P D',
 'P E',
 'P F',
 'P G',
 'P H',
 'P I',
 'P J',
 'P K',
 'P L',
 'P M',
 'P N',
 'P O',
 'P P',
 'P Q',
 'P R',
 'P S',
 'P T',
 'P U',
 'P V',
 'P W',
 'P X',
 'P Y',
 'P Z',
 'Q A',
 'Q B',
 'Q C',
 'Q D',
 'Q E',
 'Q F',
 'Q G',
 'Q H',
 'Q I',
 'Q J',
 'Q K',
 'Q L',
 'Q M',
 'Q N',
 'Q O',
 'Q P',
 'Q Q',
 'Q R',
 'Q S',
 'Q T',
 'Q U',
 'Q V',
 'Q W',
 'Q X',
 'Q Y',
 'Q Z',
 'R A',
 'R B',
 'R C',
 'R D',
 'R E',
 'R F',
 'R G',
 'R H',
 'R I',
 'R J',
 'R K',
 'R L',
 'R M',
 'R N',
 'R O',
 'R P',
 'R Q',
 'R R',
 'R S',
 'R T',
 'R U',
 'R V',
 'R W',
 'R X',
 'R Y',
 'R Z',
 'S A',
 'S B',
 'S C',
 'S D',
 'S E',
 'S F',
 'S G',
 'S H',
 'S I',
 'S J',
 'S K',
 'S L',
 'S M',
 'S N',
 'S O',
 'S P',
 'S Q',
 'S R',
 'S S',
 'S T',
 'S U',
 'S V',
 'S W',
 'S X',
 'S Y',
 'S Z',
 'T A',
 'T B',
 'T C',
 'T D',
 'T E',
 'T F',
 'T G',
 'T H',
 'T I',
 'T J',
 'T K',
 'T L',
 'T M',
 'T N',
 'T O',
 'T P',
 'T Q',
 'T R',
 'T S',
 'T T',
 'T U',
 'T V',
 'T W',
 'T X',


## Plumbing Data (Date Data of Extraction: `2 Sep 2021`)

In [8]:
driver = webdriver.Chrome()

In [9]:
all_licensee_names=[]
all_licensee_detail=[]

for i in Letters_comb[393:]:
    driver.get('https://consumer.etoolbox.pic.vic.gov.au/_layouts/cc/pic_validatepp.aspx')
    
    driver.find_element_by_xpath('//*[@id="ctl00_PlaceHolderMain_uxSurname"]').send_keys(i)
    driver.find_element_by_xpath('//*[@id="ctl00_PlaceHolderMain_uxSearch"]').click()
    
    page_source = driver.page_source
    soup =  BeautifulSoup(page_source, 'lxml')
    
    licensee_names=[] 
    licensee_names = [i.text for i in soup.findAll('span', {'class': 'prac-result'})] #Getting the count of presesnt licenses for each letter combination
    #print(licensee_names)
    #len(licensee_names)
    all_licensee_names=all_licensee_names+licensee_names
    #print(all_licensee_names)
    
    xpaths=[]
    for j in range(1,len(licensee_names)+1): #creating xpaths: just change span[i] to click different links
        xpaths.append('//*[@id="aspnetForm"]/div[3]/table/tbody/tr[1]/td[2]/table/tbody/tr/td[1]/table/tbody/tr[5]/td/div/span['+str(j) + ']')
    #print(xpaths)  
    
    licensee_detail=[]
    for k in range(1,len(licensee_names)+1): #getting the details of every license in a combination
        try:
            driver.find_element_by_xpath(xpaths[k-1]).click()
            time.sleep(1)
    
            page_source = driver.page_source
            soup =  BeautifulSoup(page_source, 'lxml')
            
    
            licensee_detail.append([i.text for i in soup.findAll('table', {'id': 'tblRegSummary'})]) 
        
        except:
            print("No data for {}".format(i))
    
    #print(licensee_detail)  
    #len(licensee_detail)     
    all_licensee_detail=all_licensee_detail+licensee_detail
    #print(all_licensee_detail)
    

In [11]:
all_licensee_names

['100873 - Peter Bowd',
 '101382 - Peter Wakefield',
 '101659 - Peter Mold',
 '101950 - Patrick Angland',
 '102045 - Paul Howland',
 '102164 - Paul Archibald',
 '102945 - Patrick Kermond',
 '103163 - Paul Howard',
 '104184 - Paul Gould',
 '10558 - Peter Blachford',
 '10567 - Paul Frood',
 '105694 - Patrick Allard',
 '106521 - Paul McDonald',
 '106857 - Paul Hayward',
 '107672 - Philip Hubbard',
 '108414 - Paul Richmond',
 '108858 - Peter Shepherd',
 '109186 - Paul Euvrard',
 '109287 - Paul Hooiveld',
 '109299 - Paul Sund',
 '109863 - Paul Croxford',
 '110162 - Patrick Howard',
 '110300 - Patrick Harwood',
 '112145 - Patrick Boland',
 '112953 - Paul Richmond',
 '113551 - Peyton Ellwood',
 '113610 - Praneel Prasad',
 '114653 - Paul Gould',
 '117437 - Peter Ackroyd',
 '117650 - Patrick Reid',
 '119035 - Paul Gillard',
 '120125 - Peri McDonald',
 "120305 - Paul O'Dowd",
 '120322 - Peter Eastwood',
 '121000 - Paul McRedmond',
 '13131 - Peter Orpwood',
 '14442 - Paul Pritchard',
 '15009 - Pe

In [12]:
len(all_licensee_names)

4671

In [13]:
all_licensee_detail

[['\n\n100873 - Peter Bowd\n\n\n\nWhile we have a record for Peter Bowd on our database, verification of their details cannot be provided. Please call the Commission on 1300 815 127 for further information.\n\n'],
 ['\n\n101382 - Peter Wakefield\n\n\n\nHolds a Registration accreditation in the following classes until 18/11/2018:\n\n\nMechanical services -  [Current]Mechanical services – restricted to duct fixing -  [Current]Mechanical services – restricted to single head split systems -  [Current]Mechanical services – restricted to solid fuel heaters -  [Current]Refrigerated air-conditioning – restricted to basic systems - (2008) -  [Current]Refrigerated air-conditioning – restricted to split systems -  [Current]\n\n\nHolds a Registration accreditation in the following classes until 8/12/2022:\n\n\nMechanical services -  [Current]Mechanical services – restricted to duct fixing -  [Current]Mechanical services – restricted to single head split systems -  [Current]Mechanical services – re

In [14]:
len(all_licensee_detail)

4671

In [15]:
data = {'Licensee Name':all_licensee_names, 'Detail':all_licensee_detail}

# Create DataFrame
df = pd.DataFrame(data)
 
# Print the output
df.head()

Licensee Name                                             Detail
0       100873 - Peter Bowd  [\n\n100873 - Peter Bowd\n\n\n\nWhile we have ...
1  101382 - Peter Wakefield  [\n\n101382 - Peter Wakefield\n\n\n\nHolds a R...
2       101659 - Peter Mold  [\n\n101659 - Peter Mold\n\n\n\nHolds a Licenc...
3  101950 - Patrick Angland  [\n\n101950 - Patrick Angland\n\n\n\nHolds a R...
4     102045 - Paul Howland  [\n\n102045 - Paul Howland\n\n\n\nWhile we hav...

In [16]:
df.tail()

Licensee Name  \
4666       117305 - Zachary Barry   
4667         121106 - Zac Lindsey   
4668  111171 - Zachariah Gomularz   
4669         113922 - Zac Jimenez   
4670            51298 - Zac Papez   

                                                 Detail  
4666  [\n\n117305 - Zachary Barry\n\n\n\nHolds a Reg...  
4667  [\n\n121106 - Zac Lindsey\n\n\n\nHolds a Regis...  
4668  [\n\n111171 - Zachariah Gomularz\n\n\n\nHolds ...  
4669  [\n\n113922 - Zac Jimenez\n\n\n\nWhile we have...  
4670  [\n\n51298 - Zac Papez\n\n\n\nWhile we have a ...

In [17]:
df.to_csv('Vic_Plumbing_final_Part_3.csv')

## Concacating multiple .csv files to get a final one

In [10]:
df1 = pd.read_csv("/Users/rohit_manral/Webscrape_Udemy_1/My_work/Vic_Plumbing_final_Part_1.csv")
df1.head()

Unnamed: 0              Licensee Name  \
0           0     100612 - Andrew Saliba   
1           1    100639 - Adam Pocervina   
2           2     101365 - Albert Sotara   
3           3  101450 - Anthony Van Osta   
4           4       101495 - Abdul Yehia   

                                              Detail  
0  ['\n\n100612 - Andrew Saliba\n\n\n\nHolds a Li...  
1  ['\n\n100639 - Adam Pocervina\n\n\n\nHolds a R...  
2  ['\n\n101365 - Albert Sotara\n\n\n\nWhile we h...  
3  ['\n\n101450 - Anthony Van Osta\n\n\n\nHolds a...  
4  ['\n\n101495 - Abdul Yehia\n\n\n\nHolds a Lice...

In [11]:
df2 = pd.read_csv("/Users/rohit_manral/Webscrape_Udemy_1/My_work/Vic_Plumbing_final_Part_2.csv")
df2.head()

Unnamed: 0           Licensee Name  \
0           0     100068 - Grant East   
1           1   100186 - Gregory Best   
2           2   101339 - Gerard Scott   
3           3  101861 - Glenn Wishart   
4           4  102190 - Garry Targett   

                                              Detail  
0  ['\n\n100068 - Grant East\n\n\n\nHolds a Regis...  
1  ['\n\n100186 - Gregory Best\n\n\n\nWhile we ha...  
2  ['\n\n101339 - Gerard Scott\n\n\n\nWhile we ha...  
3  ['\n\n101861 - Glenn Wishart\n\n\n\nWhile we h...  
4  ['\n\n102190 - Garry Targett\n\n\n\nHolds a Re...

In [12]:
df3 = pd.read_csv("/Users/rohit_manral/Webscrape_Udemy_1/My_work/Vic_Plumbing_final_Part_3.csv")
df3.head()

Unnamed: 0             Licensee Name  \
0           0       100873 - Peter Bowd   
1           1  101382 - Peter Wakefield   
2           2       101659 - Peter Mold   
3           3  101950 - Patrick Angland   
4           4     102045 - Paul Howland   

                                              Detail  
0  ['\n\n100873 - Peter Bowd\n\n\n\nWhile we have...  
1  ['\n\n101382 - Peter Wakefield\n\n\n\nHolds a ...  
2  ['\n\n101659 - Peter Mold\n\n\n\nHolds a Licen...  
3  ['\n\n101950 - Patrick Angland\n\n\n\nHolds a ...  
4  ['\n\n102045 - Paul Howland\n\n\n\nWhile we ha...

In [13]:
df = pd.concat([df1, df2, df3])
df.head()

Unnamed: 0              Licensee Name  \
0           0     100612 - Andrew Saliba   
1           1    100639 - Adam Pocervina   
2           2     101365 - Albert Sotara   
3           3  101450 - Anthony Van Osta   
4           4       101495 - Abdul Yehia   

                                              Detail  
0  ['\n\n100612 - Andrew Saliba\n\n\n\nHolds a Li...  
1  ['\n\n100639 - Adam Pocervina\n\n\n\nHolds a R...  
2  ['\n\n101365 - Albert Sotara\n\n\n\nWhile we h...  
3  ['\n\n101450 - Anthony Van Osta\n\n\n\nHolds a...  
4  ['\n\n101495 - Abdul Yehia\n\n\n\nHolds a Lice...

In [14]:
df.columns

Index(['Unnamed: 0', 'Licensee Name', 'Detail'], dtype='object')

In [15]:
df = df.drop('Unnamed: 0', 1)
df.head()

Licensee Name  \
0     100612 - Andrew Saliba   
1    100639 - Adam Pocervina   
2     101365 - Albert Sotara   
3  101450 - Anthony Van Osta   
4       101495 - Abdul Yehia   

                                              Detail  
0  ['\n\n100612 - Andrew Saliba\n\n\n\nHolds a Li...  
1  ['\n\n100639 - Adam Pocervina\n\n\n\nHolds a R...  
2  ['\n\n101365 - Albert Sotara\n\n\n\nWhile we h...  
3  ['\n\n101450 - Anthony Van Osta\n\n\n\nHolds a...  
4  ['\n\n101495 - Abdul Yehia\n\n\n\nHolds a Lice...

In [18]:
print(df.shape)

(15776, 2)


#### **Our dataset got 15776 rows and 2 columns**

# **Step 2: `Data Cleaning`**

* **Removing Duplicates**: Removing rows with same `Licensee Name` 

In [19]:
# dropping ALL duplicate values
df.drop_duplicates(subset ="Licensee Name",
                     keep = False, inplace = True)
 
print(df.shape)

(15658, 2)


In [27]:
df.to_csv('Vic_Plumbing_uncleaned_data.csv')

In [3]:
df = pd.read_csv("/Users/rohit_manral/Webscrape_Udemy_1/My_work/Vic_Plumbing_uncleaned_data.csv")
df.head()

Unnamed: 0              Licensee Name  \
0           0     100612 - Andrew Saliba   
1           1    100639 - Adam Pocervina   
2           2     101365 - Albert Sotara   
3           3  101450 - Anthony Van Osta   
4           4       101495 - Abdul Yehia   

                                              Detail  
0  ['\n\n100612 - Andrew Saliba\n\n\n\nHolds a Li...  
1  ['\n\n100639 - Adam Pocervina\n\n\n\nHolds a R...  
2  ['\n\n101365 - Albert Sotara\n\n\n\nWhile we h...  
3  ['\n\n101450 - Anthony Van Osta\n\n\n\nHolds a...  
4  ['\n\n101495 - Abdul Yehia\n\n\n\nHolds a Lice...

In [4]:
df = df.drop('Unnamed: 0', 1)
df.head()

Licensee Name  \
0     100612 - Andrew Saliba   
1    100639 - Adam Pocervina   
2     101365 - Albert Sotara   
3  101450 - Anthony Van Osta   
4       101495 - Abdul Yehia   

                                              Detail  
0  ['\n\n100612 - Andrew Saliba\n\n\n\nHolds a Li...  
1  ['\n\n100639 - Adam Pocervina\n\n\n\nHolds a R...  
2  ['\n\n101365 - Albert Sotara\n\n\n\nWhile we h...  
3  ['\n\n101450 - Anthony Van Osta\n\n\n\nHolds a...  
4  ['\n\n101495 - Abdul Yehia\n\n\n\nHolds a Lice...

#### **Our final dataset got 15658 rows and 2 columns**

* **Retrieving Licensee Name**

In [6]:
names=[]
for i in df['Licensee Name']:
    names.append(i.split('-')[1].strip()) 
    
names

['Andrew Saliba',
 'Adam Pocervina',
 'Albert Sotara',
 'Anthony Van Osta',
 'Abdul Yehia',
 'Anthony Van Osta',
 'Andrew Corica',
 'Ali Kara',
 'Adam Jana',
 'Adam Zerella',
 'Aydin Balla',
 'Amit Chopra',
 'Ahmet Jeka',
 'Aleksandar Trkulja',
 'Ahmad Chamra',
 'Antonio Muraca',
 'Aaron Niezgodha',
 'Aymon Rozema',
 'Aleksandar Suvira',
 'Alex Weetra',
 'Alexander Kellalea',
 'Armando De Oliveira',
 "Andrew O'Dea",
 'Adikaramage Jayawardena',
 'Aaron Niezgodka',
 'Alexander Weetra',
 'Anthony Cappadona',
 'Ahmed Reza Topiwalla',
 'Arianit Meta',
 'Ankido Dankha',
 'Adam Kinsella',
 'Adam Lamanna',
 'Arnol Abarca',
 'Alexander Fusca',
 'Alastair Kosta',
 'Ahmad Nasra',
 'Anthony Caruana',
 'Anto Kustura',
 'Alofa Mika',
 'Andrew Celea',
 'Andrew Bracka',
 'Anthony La Spina',
 'Antony Sicura',
 'Alex Penaluna',
 'Asanka Kulatunga',
 'Adam Zanella',
 'Andrew Luxa',
 'Adam Pannia',
 'Adrian Sacchetta',
 'Anil Weerasena',
 'Andrew Habib',
 'Adam Roob',
 'Abad Al Nosir Taleb',
 'Adam Chehab

* **Retrieving License Number**

In [5]:
num=[]
for i in df['Licensee Name']:
    num.append(i.split('-')[0].strip()) 
    
num

['100612',
 '100639',
 '101365',
 '101450',
 '101495',
 '101613',
 '101816',
 '103198',
 '103502',
 '103993',
 '104313',
 '104552',
 '104571',
 '104652',
 '105050',
 '105183',
 '105734',
 '105824',
 '105978',
 '106031',
 '106147',
 '106225',
 '106238',
 '106332',
 '106398',
 '106544',
 '106694',
 '108107',
 '108289',
 '108327',
 '108351',
 '108391',
 '108437',
 '108637',
 '108791',
 '108926',
 '109154',
 '109175',
 '109744',
 '109784',
 '110039',
 '110464',
 '110607',
 '110611',
 '110909',
 '110970',
 '111166',
 '111169',
 '111259',
 '111309',
 '100135',
 '100145',
 '101473',
 '101521',
 '102820',
 '103405',
 '10505',
 '108784',
 '111988',
 '114272',
 '114411',
 '114573',
 '116565',
 '119365',
 '31188',
 '35376',
 '39109',
 '41232',
 '44974',
 '49420',
 '49707',
 '49750',
 '53274',
 '8573',
 '101305',
 '103692',
 '104047',
 '104465',
 '105504',
 '105772',
 '106952',
 '107361',
 '108705',
 '109328',
 '109485',
 '110943',
 '111217',
 '111528',
 '112109',
 '112233',
 '114467',
 '114468',


* **Retrieving License Status**

In [33]:
stat=[]
for i in df['Detail']:
    if (i.find("[Current]") == -1):
        stat.append("Inactive") 
    else:
        stat.append("Active") 
    
stat

['Active',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Inactive',
 'Active',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Active',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Inactive',
 'Inactive',
 'Inactive',
 'Active',
 'Active',
 'Inactive',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Active',
 'Inactive',
 'Inactive'

* **Retrieving Expiry Date**

* **Retrieving License Class(es)**

In [ ]:
classes=[]
for i in df['Detail']:
    
    
classes